### PR TITLE
object: remove default value us-east-1 as region from s3 client

### PR DESCRIPTION
**Pull request by thotz**
Currently s3 client apis sets default value as `us-east-1`. This patch
removes that, region need to map with zonegroup in the RGW server so we
set that accordingly

Signed-off-by: Jiffin Tony Thottan <thottanjiffin@gmail.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: tests/framework/clients/bucket.go [hunk 1]
```diff
@@ -157,9 +157,9 @@ func (b *BucketOperation) CheckBucketNotificationSetonRGW(namespace, storeName,
 	s3AccessKey, _ := helper.BucketClient.GetAccessKey(obcName)
 	s3SecretKey, _ := helper.BucketClient.GetSecretKey(obcName)
 	if tlsEnabled {
-		s3client, err = rgw.NewInsecureS3Agent(s3AccessKey, s3SecretKey, s3endpoint, "", true)
+		s3client, err = rgw.NewInsecureS3Agent(s3AccessKey, s3SecretKey, s3endpoint, storeName, true)
```
Comment (travisn) [inline tests/framework/clients/bucket.go:160]: If a different region is specified in the storage class, will this client connection work with the store name? Or does it need to match the storage class?

File: pkg/operator/ceph/object/bucket/provisioner.go [hunk 2]
```diff
@@ -513,7 +513,11 @@ func (p *Provisioner) setEndpoint(sc *storagev1.StorageClass) {
 
 func (p *Provisioner) setRegion(sc *storagev1.StorageClass) {
 	const key = "region"
-	p.region = sc.Parameters[key]
+	if len(sc.Parameters[key]) > 0 {
+		p.region = sc.Parameters[key]
+	} else {
+		p.region = p.objectContext.ZoneGroup
```
Comment (leseb) [inline pkg/operator/ceph/object/bucket/provisioner.go:519]: Can we understand in a comment, why we assign ZoneGroup and not Realm or Zone?

**Review by leseb — CHANGES_REQUESTED** [review]

File: tests/framework/clients/bucket.go (hunk 1 above)
Comment (thotz) [reply tests/framework/clients/bucket.go:160]: > `us-east-1` is still used in the integration tests in ceph_base_object_test.go. How about changing that to something else so we can see the non-default works?
@travisn : do u want to check whether with regions and see s3 client works or not right? It more or like s3 client kind of test, but I can definitely add those

**Comment by thotz** [comment]
> `us-east-1` is still used in the integration tests in ceph_base_object_test.go. How about changing that to something else so we can see the non-default works?

@travisn : do u want to check whether with regions and see s3 client works or not right? It's more or like s3 client kind of test, but I can definitely add those

File: deploy/examples/storageclass-bucket-delete.yaml [hunk 3]
```diff
@@ -9,7 +9,7 @@ reclaimPolicy: Delete
 parameters:
    objectStoreName: my-store
    objectStoreNamespace: rook-ceph # namespace:cluster
-   region: us-east-1
+   # region: us-east-1
```
[the remark this answers is not in the document]
Comment (thotz) [reply deploy/examples/storageclass-bucket-delete.yaml:12]: @BlaineEXE : I will try to add #9556 with this change

File: tests/framework/clients/bucket.go (hunk 1 above)
Comment (thotz) [reply tests/framework/clients/bucket.go:160]: @travisn : I wondered magic behind `us-east-1` as region value. The actual value for s3 client needs to `storeName` and ideally we should have filled same value for storage class. The `us-east-1` works for RGW, I am guessing it skips checks(or maybe bug)

**Comment by travisn** [comment]
> > `us-east-1` is still used in the integration tests in ceph_base_object_test.go. How about changing that to something else so we can see the non-default works?
> 
> @travisn : do u want to check whether with regions and see s3 client works or not right? It's more or like s3 client kind of test, but I can definitely add those

Yes sounds to add some testing for this.

File: tests/framework/clients/bucket.go (hunk 1 above)
Comment (thotz) [reply tests/framework/clients/bucket.go:160]: @travisn : The only `create bucket` request needs the correct region value, all the objects operation works fine with the 
invalid region for s3 client. So I guess I need to add a test case for `cephobjectstoreuser` with creating a bucket to test this out

[Another update]: `us-east-1` as a region no longer valid in master, it works with pacific though. So IMO we might seen similar issue when we start using new ceph versions

**Comment by thotz** [comment]
> > > `us-east-1` is still used in the integration tests in ceph_base_object_test.go. How about changing that to something else so we can see the non-default works?
> > 
> > 
> > @travisn : do u want to check whether with regions and see s3 client works or not right? It's more or like s3 client kind of test, but I can definitely add those
> 
> Yes sounds to add some testing for this.

The only `create bucket` request needs the correct region value, all the objects operation works fine with the 
invalid region for s3 client. So I guess I need to add a test case for `cephobjectstoreuser` by creating a bucket to test this out

File: Documentation/ceph-object-bucket-claim.md [hunk 4]
```diff
@@ -105,13 +105,13 @@ provisioner: rook-ceph.ceph.rook.io/bucket [2]
...
-  region: us-west-1
-  bucketName: ceph-bucket [4]
-reclaimPolicy: Delete [5]
+  region: us-west-1 [4]
+  bucketName: ceph-bucket [5]
+reclaimPolicy: Delete [6]
 ```
 1. `label`(optional) here associates this `StorageClass` to a specific provisioner.
 1. `provisioner` responsible for handling `OBCs` referencing this `StorageClass`.
-1. **all** `parameter` required.
```
Comment (travisn) [inline Documentation/ceph-object-bucket-claim.md:114]: If this line is removed, the numbering is off. I agree this bullet didn't add value, so you just need to decrement the references for 4, 5, and 6 above.

File: Documentation/ceph-object-bucket-claim.md [hunk 5]
```diff
@@ -105,13 +105,13 @@ provisioner: rook-ceph.ceph.rook.io/bucket [2]
...
+1. `region`(optional) defines the region in which the bucket should be created. For RGW, this maps to `zonegroup` of the server. If user does not define `region`, Rook will fill this from in-memory `objectStoreContext` for the `ceph object store` and by default this value is the name of `ceph object store` for the internal cluster. If region is specified by the user and if it is different from default value, then create [CephObjectZoneGroup](ceph-object-multisite-crd.md#object-zone-group-settings) for the `ceph object store`. Please refer to [Ceph's documentation](https://docs.ceph.com/en/latest/radosgw/multisite/#zone-groups) for more details.
```
Comment (travisn) [inline Documentation/ceph-object-bucket-claim.md:114]: Can we shorten this description? There is a lot of detail the user doesn't need to know, such as "in-memory objectStoreContext". All they need to know is if they want to set a region name, what should their settings be in the CRs? I'm not even clear on how to do it. We might need two separate examples for 1) the simple local object store and 2) the multisite.

File: tests/framework/clients/bucket.go (hunk 1 above)
Comment (travisn) [reply tests/framework/clients/bucket.go:160]: So if we remove region from the default examples in rook, it will work with ceph pacific and master? Or will we need different examples for pacific vs quincy? And to be clear, does the ceph master change also affect quincy, or just post-quincy?

**Review by travisn — CHANGES_REQUESTED** [review]

File: tests/framework/clients/bucket.go (hunk 1 above)
Comment (thotz) [reply tests/framework/clients/bucket.go:160]: Yeah, it will work, this fills internal by Rook. Earlier we are using `us-east-1` from this patch we add the actual zone of the rgw server that's the difference.
The `us-east-1` worked in quincy as well, but I don't see much code difference in that path for master and Quincy. Need to have a bit of time to understand that

File: Documentation/ceph-object-bucket-claim.md [hunk 6]
```diff
@@ -102,18 +102,92 @@ metadata:
...
-1. **all** `parameter` required.
+1. `region`(optional) defines the region in which the bucket should be created. For RGW, this maps to [zonegroup](https://docs.ceph.com/en/latest/radosgw/multisite/#zone-groups) of the server. If user does not define `region`, Rook will fill this value accordingly. For examples please check at the end of the doc.
 1. `bucketName` is required for access to existing buckets but is omitted when provisioning new buckets.
 Unlike greenfield provisioning, the brownfield bucket name appears in the `StorageClass`, not the `OBC`.
 1. rook-ceph provisioner decides how to treat the `reclaimPolicy` when an `OBC` is deleted for the bucket. See explanation as [specified in Kubernetes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#retain)
 + _Delete_ = physically delete the bucket.
 + _Retain_ = do not physically delete the bucket.
+
+
+##### Storage Class for simple local object store
```
Comment (travisn) [inline Documentation/ceph-object-bucket-claim.md:122]: How about this title? Also it would help to describe the example briefly and explain what the default zone will be. "my-store" in this case?
```suggestion
##### Example: Default region
```

File: Documentation/ceph-object-bucket-claim.md [hunk 7]
```diff
@@ -102,18 +102,92 @@ metadata:
...
+metadata:
+   name: rook-ceph-delete-bucket
+provisioner: rook-ceph.ceph.rook.io/bucket
+reclaimPolicy: Delete
+parameters:
+   objectStoreName: my-store
+   objectStoreNamespace: rook-ceph
+```
+
+##### CRDs for Storage Class, if region defined
```
Comment (travisn) [inline Documentation/ceph-object-bucket-claim.md:135]: A brief description of the example would also help here. You might also consider cutting out sections of the yaml that is not needed for the example, just to show the properties that must be set for a custom region.
```suggestion
##### Example: Custom region
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: tests/framework/clients/bucket.go (hunk 1 above)
Comment (thotz) [reply tests/framework/clients/bucket.go:160]: I need to take back about `us-east-1` region value, this will work with the master and Quincy branch. Earlier it was a typo mistake from my end

File: Documentation/ceph-object-bucket-claim.md (hunk 7 above)
Comment (thotz) [reply Documentation/ceph-object-bucket-claim.md:135]: @travisn : I am not sure what part can be cut off, for the region, a `Zonegroup` need to define. In the RGW use cases most like to have zone and zone group coexists. Rook mandates realm as part of `ZoneGroup` CRD. So in short with the current implementation users need to define all three.

File: Documentation/ceph-object-bucket-claim.md [hunk 8]
```diff
@@ -102,18 +102,94 @@ metadata:
...
+1. `region`(optional) defines the region in which the bucket should be created. For RGW, this maps to [zonegroup](https://docs.ceph.com/en/latest/radosgw/multisite/#zone-groups) of the server. If user does not define `region`, Rook will fill this value accordingly. For examples please check at the end of the doc.
 1. `bucketName` is required for access to existing buckets but is omitted when provisioning new buckets.
 Unlike greenfield provisioning, the brownfield bucket name appears in the `StorageClass`, not the `OBC`.
 1. rook-ceph provisioner decides how to treat the `reclaimPolicy` when an `OBC` is deleted for the bucket. See explanation as [specified in Kubernetes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#retain)
 + _Delete_ = physically delete the bucket.
 + _Retain_ = do not physically delete the bucket.
+
+
+##### Example: Default region
+In the below example `region` is not defined, so Rook Operator will fetch `zonegroup` from the `RGW` server and fill this out.
```
Comment (travisn) [inline Documentation/ceph-object-bucket-claim.md:123]: Can you add what the region is in this example? my-store?
```suggestion
In the below example `region` is not defined, so Rook Operator will default to the name of the `zonegroup`, which is the same as the object store name `my-store`. 
```

File: Documentation/ceph-object-bucket-claim.md [hunk 9]
```diff
@@ -102,18 +102,94 @@ metadata:
...
+   name: rook-ceph-delete-bucket
+provisioner: rook-ceph.ceph.rook.io/bucket
+reclaimPolicy: Delete
+parameters:
+   objectStoreName: my-store
+   objectStoreNamespace: rook-ceph
+```
+
+##### Example: Custom Region
+Here user is providing a custom value for the `region`. So make sure following [multisite related CRDs](ceph-object-multisite-crd.md) are defined as well. The Rook Operator will create required `zonegroup` for RGW server.
```
Comment (travisn) [inline Documentation/ceph-object-bucket-claim.md:137]:
```suggestion
Here user is providing a custom value `us-west-1` for the `region`. So make sure following [multisite related CRDs](ceph-object-multisite-crd.md) are defined as well. The Rook Operator will create required `zonegroup` for RGW server.
```

File: tests/framework/clients/bucket.go (hunk 1 above)
Comment (travisn) [reply tests/framework/clients/bucket.go:160]: The documented examples we are working on in the other comments only apply to Quincy and newer, correct? So what is the behavior for region in Pacific? The docs need to give that detail, especially since Pacific is still the default version with Rook.

**Review by travisn — CHANGES_REQUESTED** [review]

File: Documentation/ceph-object-bucket-claim.md (hunk 8 above)
Comment (thotz) [reply Documentation/ceph-object-bucket-claim.md:123]: @travisn: This is not completely correct, zonegroup in object context can have the following values.
1.) normal simple cluster -> name of the object-store
2.) external cluster -> `default` 
3.) if multisite configure then it will be value created by the user
Please check `getMultisiteForObjectStore()` for more details. In any of the cases, if the user does not define region value, the Rook operator will fetch the value from `object Store Context` and set it accordingly. Please note if the user provides custom `region` then he configure multisite settings but if RGW configured with multisite it does not mean user need to provide the `region` value

Comment (travisn) [reply Documentation/ceph-object-bucket-claim.md:123]: > @travisn: This is not completely correct, zonegroup in object context can have the following values. 
> 1.) normal simple cluster -> name of the object-store 

Ok, this is what I expect this first example to be... the local simple cluster. 

> 2.) external cluster -> `default` 

Perhaps we need an external cluster example that points out it could be anything, depending on the external cluster config.

> 3.) if multisite configure then it will be value created by the user 
> Please check `getMultisiteForObjectStore()` for more details. In any of the cases, if the user does not define region value, the Rook operator will fetch the value from `object Store Context` and set it accordingly. Please note if the user provides custom `region` then he configure multisite settings but if RGW configured with multisite it does not mean user need to provide the `region` value

This case I thought was the second example below. If it's not just the zonegroup name, then it's very confusing to me. We need to figure out how to simplify the explanation in the docs, we can't ask the users to look at the code.

File: tests/framework/clients/bucket.go (hunk 1 above)
Comment (travisn) [reply tests/framework/clients/bucket.go:160]: I'm still confused about how users can understand the region. If `us-east-1` is a workaround and a hack, maybe we should just keep that for the simple, local rgw case. What does the user gain by using a different region? Seems like it should only be important for external or multi-site where we need the region name in order to connect. But if it's just the local cluster, can we hide the region from the user so they don't need to worry about it?

Comment (travisn) [reply tests/framework/clients/bucket.go:160]: Let's find a time to discuss this next week since I'm having such a hard time. :)

Comment (thotz) [reply tests/framework/clients/bucket.go:160]: @travisn: Sure :). Sounds good to me

File: Documentation/ceph-object-bucket-claim.md (hunk 8 above)
Comment (travisn) [reply Documentation/ceph-object-bucket-claim.md:123]: Could we rewrite the new documentation for `region` and the examples to be very clear on these two points? I'm still not clear on these:
1. us-east-1 will always work for their region 
2. They can use a different region by following the docs in this PR, but it is totally optional

Comment (thotz) [reply Documentation/ceph-object-bucket-claim.md:123]: > Could we rewrite the new documentation for `region` and the examples to be very clear on these two points? I'm still not clear on these:
> 
>     1. us-east-1 will always work for their region
> 
@travisn I have mentioned it in [second example](https://github.com/rook/rook/pull/9754/files#diff-fb33f0fa41ca6346e7562087d2da05cc32fb65ea2587b23bf36b866a9a609720R137) and thought of keeping the first example leaving the `region` value. So is it better to modify the first one to include `us-east-1` than leaving out the `region` value?

>     2. They can use a different region by following the docs in this PR, but it is totally optional

Comment (travisn) [reply Documentation/ceph-object-bucket-claim.md:123]: Maybe we can just remove the first example. It seems sufficient that the second example mentions the default is `us-east-1`, and if they want to change the region, this is how they do it.

Comment (thotz) [reply Documentation/ceph-object-bucket-claim.md]: @travisn: Thanks. I have updated the PR accordingly

File: Documentation/ceph-object-bucket-claim.md [hunk 10]
```diff
@@ -102,18 +102,79 @@ metadata:
...
   objectStoreNamespace: rook-ceph
-  region: us-west-1
+  region: us-west-1 [3]
   bucketName: ceph-bucket [4]
 reclaimPolicy: Delete [5]
 ```
 1. `label`(optional) here associates this `StorageClass` to a specific provisioner.
 1. `provisioner` responsible for handling `OBCs` referencing this `StorageClass`.
-1. **all** `parameter` required.
+1. `region`(optional) defines the region in which the bucket should be created. For RGW, this maps to [zonegroup](https://docs.ceph.com/en/latest/radosgw/multisite/#zone-groups) of the server. If user does not define `region`, Rook will fill this value accordingly. Please check example at the end of the doc.
```
Comment (travisn) [inline Documentation/ceph-object-bucket-claim.md:114]: Shall we mention the default here? Otherwise, I'm not sure what it means for Rook to "fill this value".
```suggestion
1. `region`(optional) defines the region in which the bucket should be created. For RGW, this maps to [zonegroup](https://docs.ceph.com/en/latest/radosgw/multisite/#zone-groups) of the server. If user does not define `region`, the default is `us-east-1`. Please check example at the end of the doc.
```

File: deploy/examples/storageclass-bucket-delete.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (travisn) [reply deploy/examples/storageclass-bucket-delete.yaml:12]: @thotz Since the default is `us-east-1`, should we just leave this default uncommented?
```suggestion
   region: us-east-1
```

File: deploy/examples/storageclass-bucket-retain.yaml [hunk 11]
```diff
@@ -8,7 +8,7 @@ reclaimPolicy: Retain
 parameters:
    objectStoreName: my-store # port 80 assumed
    objectStoreNamespace: rook-ceph # namespace:cluster
-   region: us-east-1
+   # region: us-east-1
```
Comment (travisn) [inline deploy/examples/storageclass-bucket-retain.yaml:11]:
```suggestion
   region: us-east-1
```

File: tests/framework/clients/bucket.go [hunk 12]
```diff
@@ -157,9 +157,9 @@ func (b *BucketOperation) CheckBucketNotificationSetonRGW(namespace, storeName,
...
 	} else {
-		s3client, err = rgw.NewS3Agent(s3AccessKey, s3SecretKey, s3endpoint, "", true, nil)
+		s3client, err = rgw.NewS3Agent(s3AccessKey, s3SecretKey, s3endpoint, storeName, true, nil)
```
Comment (travisn) [inline tests/framework/clients/bucket.go:162]: Why is the storeName used here, but the zonegroup is used in the example for using a different region? Shouldn't this use the zonegroup as well? Or in the test, the storeName and zonegroup are the same anyway?

File: Documentation/ceph-object-bucket-claim.md [hunk 13]
```diff
@@ -102,18 +102,79 @@ metadata:
...
 1. `bucketName` is required for access to existing buckets but is omitted when provisioning new buckets.
 Unlike greenfield provisioning, the brownfield bucket name appears in the `StorageClass`, not the `OBC`.
 1. rook-ceph provisioner decides how to treat the `reclaimPolicy` when an `OBC` is deleted for the bucket. See explanation as [specified in Kubernetes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#retain)
 + _Delete_ = physically delete the bucket.
 + _Retain_ = do not physically delete the bucket.
+
+##### Example: Custom Region
+If user is providing the `region` value other than `us-east-1` then additional CRDs for zonegroup/zone/realm need to be created. For example, here user is providing `region` value as `us-west-1`. So make sure following [multisite related CRDs](ceph-object-multisite-crd.md) are defined as well. The Rook Operator will create required `zonegroup` for RGW server.
```
Comment (travisn) [inline Documentation/ceph-object-bucket-claim.md:122]:
```suggestion
To specify a `region` other than `us-east-1`, a custom zone group must be created that matches the desired region name. In the following example, the `region` is set to `us-west-1`. For more details, see the [related multisite CRDs](ceph-object-multisite-crd.md). 
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/object/bucket/provisioner.go [hunk 14]
```diff
@@ -513,7 +513,13 @@ func (p *Provisioner) setEndpoint(sc *storagev1.StorageClass) {
...
 func (p *Provisioner) setRegion(sc *storagev1.StorageClass) {
 	const key = "region"
-	p.region = sc.Parameters[key]
+	if len(sc.Parameters[key]) > 0 {
+		p.region = sc.Parameters[key]
+	} else {
+		// If user does not define, then set region to Zonegroup, since RGW internally maps
+		// aws region to ZoneGroup
+		p.region = p.objectContext.ZoneGroup
+	}
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/bucket/provisioner.go:522]: I'm a little bit confused. What is the point of overriding `region` if the default is mapped to `ZoneGroup` and if the user has to create a ZoneGroup resource if they want to override region? Would it be simpler to just remove this parameter and rely on users to create a `ZoneGroup` instead?

File: Documentation/ceph-object-bucket-claim.md (hunk 10 above)
Comment (BlaineEXE) [inline Documentation/ceph-object-bucket-claim.md:114]:
```suggestion
1. `region`(optional) defines the region in which the bucket should be created. For RGW, this maps to the [zone group](https://docs.ceph.com/en/latest/radosgw/multisite/#zone-groups) of the server. If user does not define `region`, Rook will fill this value accordingly. Please check example at the end of the doc.
```

Comment (BlaineEXE) [inline Documentation/ceph-object-bucket-claim.md:114]: This is still not very clear. What is the region set to if the user doesn't define it? "Accordingly" isn't clear enough.

Comment (BlaineEXE) [inline Documentation/ceph-object-bucket-claim.md:114]: Please add a link to the example in the doc for users.

File: Documentation/ceph-object-bucket-claim.md (hunk 13 above)
Comment (BlaineEXE) [inline Documentation/ceph-object-bucket-claim.md:122]: Do users need to create all of the resources, or only zone group?

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
If we can fill in a default value for "region" from the zone group, is region even a valid parameter?

**Comment by alimaredia** [comment]
@thotz I was taking at glance at this PR and I don't quite understand the use case/problem this is trying to solve, even after reading the doc changes. Could you summarize what the object-store/obc's couldn't do before this PR was written and what the PR solves?

File: Documentation/ceph-object-bucket-claim.md (hunk 10 above)
Comment (thotz) [reply Documentation/ceph-object-bucket-claim.md:114]: @travisn : I don't what I missed earlier if the user does not define the region, Rook will fill it from `Zonegroup` from `ceph object store context` and don't use `us-east-1` at all(at least that was my first intention of PR).

`us-east-1` is not the default value from RGW(at least not in RGW code base). It is a value that always works mysteriously. Maybe because a lot of s3 clients use this as default for `region`, but again not defined as default by RGW

File: pkg/operator/ceph/object/bucket/provisioner.go (hunk 14 above)
Comment (thotz) [reply pkg/operator/ceph/object/bucket/provisioner.go:522]: @BlaineEXE : `region` is an optional value for storage-class. If it is not defined then we need to fill up with `Zonegroup` from object store context, This value changes in different scenarios: 
1.) normal cluster, it is name of object store
2.) external cluster it is `default`
3.) if multiple zones are configured then it depends on value set in object-store CR.

If the user wants to use a custom region then he defines `Zonegroup` that's different scenario. Hope this avoids the confusion

File: Documentation/ceph-object-bucket-claim.md (hunk 10 above)
Comment (thotz) [reply Documentation/ceph-object-bucket-claim.md:114]: @BlaineEXE : earlier version was `Rook Operato fill this value from object store context internally`. But @travisn suggested it better not to give internal code details for the end-user. Hence changed like this. I can revert back the change

File: Documentation/ceph-object-bucket-claim.md (hunk 13 above)
Comment (thotz) [reply Documentation/ceph-object-bucket-claim.md:122]: If I understand correctly `zonegroup` is not an independent resource. Rooks mandates to have `realm` for `zonegroup` and in RGW does not make sense to have `zonegroup` without `zones` (zonegroup is collection of zones and zone is collection of RGW servers in short)

**Comment by thotz** [comment]
> @thotz I was taking at glance at this PR and I don't quite understand the use case/problem this is trying to solve, even after reading the doc changes. Could you summarize what the object-store/obc's couldn't do before this PR was written and what the PR solves?
@alimaredia : Please refer https://github.com/rook/rook/pull/9754#discussion_r821342996 for details

**Comment by thotz** [comment]
@BlaineEXE @travisn : I have minor nits need to address in this PR. But before refreshing the PR I want to make sure everyone is on the same page

File: pkg/operator/ceph/object/bucket/provisioner.go (hunk 14 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/object/bucket/provisioner.go:522]: In all three of those cases, I don't think we need to have the `region` parameter.
1. If external cluster with no zone group defined, use `default`
2. If external cluster with external zone group defined, use the zone group
3. If internal cluster with no zone group defined, use name of object store
4. If internal cluster with zone group defined, use the zone group

Is this not correct? Are there any cases missing?

File: Documentation/ceph-object-bucket-claim.md (hunk 10 above)
Comment (BlaineEXE) [reply Documentation/ceph-object-bucket-claim.md:114]: Even the text "fill this value from object store context internally" doesn't explain to the user what the value will be. In order for the user to understand when they need to modify the value, they must understand what the default value is. In words good for a user, what value will be the default?

This is also assuming that `region` is even a necessary parameter to have still.

Comment (travisn) [reply Documentation/ceph-object-bucket-claim.md:114]: Ok, does that mean rook can just take care of this internally? Why do we need to expose the `region` setting at all?

Comment (thotz) [reply Documentation/ceph-object-bucket-claim.md:114]: @travisn : The `region` defined in `storage class` of libbucket provisioner. Not sure whether Rook can hide it from user, maybe we can ignore the value. But this was available earlier release not sure whether it right to remove the variable. AFAIR I have sent https://github.com/rook/rook/pull/8766 as per request from the user, he wants to set the custom region for his use case.

File: tests/framework/clients/bucket.go (hunk 12 above)
Comment (thotz) [reply tests/framework/clients/bucket.go:162]: Yeah this should be zone group, since we testing only internal ceph cluster without multisite, I just set directly `storeName` which is value of `zonegroup` as well

File: pkg/operator/ceph/object/bucket/provisioner.go (hunk 14 above)
Comment (thotz) [reply pkg/operator/ceph/object/bucket/provisioner.go:522]: @BlaineEXE : Please see https://github.com/rook/rook/pull/9754#discussion_r822287813 why the region is needed.
Above cases are right, but please note 2.) is not support in Rook atm

File: Documentation/ceph-object-bucket-claim.md (hunk 10 above)
Comment (thotz) [reply Documentation/ceph-object-bucket-claim.md:114]: Okay will update it again next push

**Comment by thotz** [comment]
> > @BlaineEXE @travisn : I have minor nits need to address in this PR. But before refreshing the PR I want to make sure everyone is on the same page
> 
> It is still unclear what is wrong currently. The description in #9857 isn't quite enough for me to understand. Please be very clear about whether behavior is broken or not. Is there broken behavior? Can you make a test case that causes any failures? If yes, outline the failing case. If we can't make a failing test case, then why are we making the code more complex? Wouldn't that mean the parameter is just unnecessary?
> 
> Once we understand what is wrong, we can do a much better job of reviewing the PR for behavior and usability and not merely code.
There are two commits in this PR :
* replacing instance of `us-east-1` with `zone group` of RGW server
* document how to use custom `region` in `storage class` of OBC for the user

The second part is clear I guess. The reason for the why `region` is needed is explained in earlier replies

Now talking about the first part, most of the cases will work with current(i.e using `us-east-1`) but it is semantically correct to use `zone-group` instead.It is based on @cbodley's [suggestion](https://github.com/rook/rook/discussions/9142#discussioncomment-2088859).
I have added a test case with the invalid region for the s3 client https://github.com/rook/rook/pull/9754/files#diff-5f2c0f5640c927aba71e73e05b71d0f3116c972f7e0f99a40c6e2b1a986b0431R210. EIther `zonegroup` or `us-east-1` as the region for s3 client the request will succeed, so not sure whether it requires to have test cases based on it, If need I can add those as well

**Comment by BlaineEXE** [comment]
> > > @BlaineEXE @travisn : I have minor nits need to address in this PR. But before refreshing the PR I want to make sure everyone is on the same page
> > 
> > 
> > It is still unclear what is wrong currently. The description in #9857 isn't quite enough for me to understand. Please be very clear about whether behavior is broken or not. Is there broken behavior? Can you make a test case that causes any failures? If yes, outline the failing case. If we can't make a failing test case, then why are we making the code more complex? Wouldn't that mean the parameter is just unnecessary?
> > Once we understand what is wrong, we can do a much better job of reviewing the PR for behavior and usability and not merely code.
> > @BlaineEXE : there are two commits in this PR :
> 
> * replacing instance of `us-east-1` with `zone group` of RGW server
> * document how to use custom `region` in `storage class` of OBC for the user
> 
> The second part is clear I guess. The reason for the why `region` is needed is explained in earlier replies

@thotz as I explained in the message you replied to, it is not clear to those of us trying to understand what is going on. Please summarize again, and be thorough. Several of us have concerns about this approach, and we need better information to make sure we are applying the correct fix for this.

> Now talking about the first part, most of the cases will work with current(i.e using `us-east-1`) but it is semantically correct to use `zone-group` instead. It is based on @cbodley's [suggestion](https://github.com/rook/rook/discussions/9142#discussioncomment-2088859).

From that discussion, it has become clear to me that there is **something** wrong with the `region` setting, but I am not confident that the current proposed changes are the best solution. I still want some sort of proof that simply removing the field as a configurable option isn't a better solution.

> I have added a test case with the invalid region for the s3 client https://github.com/rook/rook/pull/9754/files#diff-5f2c0f5640c927aba71e73e05b71d0f3116c972f7e0f99a40c6e2b1a986b0431R210. EIther `zonegroup` or `us-east-1` as the region for s3 client the request will succeed, so not sure whether it requires to have test cases based on it, If need I can add those as well

I should have been more clear. I am not interested in seeing a unit test. Unit tests cannot show the actual behavior of a running RGW. I want to see the details for creating an end-to-end scenario where there is a failure.

From the discussion linked, it seems that we are be able to induce some failure by setting `region` to some value other than `us-east-1`. If changing this value is the only way to induce a failure, then the right solution is probably to ignore the `region` option in Rook.

In order to determine if we need changes other than removing the value entirely, we want to need to have a failing scenario with `region` left unspecified (it gets its default value). If we can make this failing scenario, then we need better defaults (but we still may not need the `region` option).

Once we have better defaults, the question becomes: can we make another failing scenario (internal or external ceph cluster) such that we can't compute the default ourselves? If we can, then users will need an option to override `region`. If we can't then the updated computation is all we needed, and we can ignore the `region` setting.

**Comment by BlaineEXE** [comment]
This documentation for RHCS 3 states that the RGW now uses the terminology "zone-group" instead of "region". https://access.redhat.com/documentation/en-us/red_hat_ceph_storage/3/html/object_gateway_guide_for_red_hat_enterprise_linux/rgw-multisite-rgw

Since Rook now has multisite configs zone/zonegroup/realm, I think we probably no longer need the "realm" option in the OBC storageclass. I think it's probably vestigial. I think that we likely only need to figure out what the zone group is and use that value.

My best **guess** at what a failing scenario would be is to :
1. Create a ceph cluster (cluster A) with an RGW using a non-"default" zone group
2. Create a Rook cluster in external mode (cluster B) whose CephObjectStore references cluster A, and do not set zone/zonegroup info.

I suspect this scenario will cause a failure because the S3 client in cluster B won't be able to connect properly to the RGW in cluster A since the zone group in cluster A is not "default". (This PR did some related work recently: https://github.com/rook/rook/pull/9751)

From the information I have, I think the most appropriate fix for this would be to make sure that creating a zone and zone group that reference the external cluster are created in cluster B. And we must also update docs and examples to remove the "realm" config.

@alimaredia has an open PR to better allow configuring zone/zonegroup endpoints for external clusters. I'm not sure if that PR is directly related to this discussion. https://github.com/rook/rook/pull/9622

@alimaredia does this breakdown make some sense to you?

**Comment by thotz** [comment]
After discussing with developers from Rook and RGW planning to close this PR infavour of https://github.com/rook/rook/pull/9906
